File: dotNetTips.Utility/Xml/XmlHelper.vb
' ***********************************************************************
' Assembly         : dotNetTips.Utility
' Author           : David McCarter
' Created          : 01-23-2016
'
' Last Modified By : David McCarter
' Last Modified On : 01-23-2016
' ***********************************************************************
' <copyright file="XmlHelper.vb" company="dotNetTips.com">
'     dotNetTips.com. All rights reserved.
' </copyright>
' <summary></summary>
' ***********************************************************************
Imports System.IO
Imports System.Xml
Imports System.Xml.Serialization

Namespace Xml
    ''' <summary>
    ''' Helper functions for XML.
    ''' </summary>
    ''' <remarks></remarks>
    Public Module XmlHelper

        ''' <summary>
        ''' Gets the node value.
        ''' </summary>
        ''' <param name="xml">The XML.</param>
        ''' <param name="path">The path.</param>
        ''' <returns></returns>
        Public Function GetNodeValue(ByVal xml As System.Xml.XPath.IXPathNavigable, ByVal path As String) As String

            If IsNothing(xml) Then
                Throw New ArgumentNullException(NameOf(xml))
            End If

            If String.IsNullOrEmpty(path) Then
                Throw New ArgumentNullException(NameOf(path))
            End If

            Try
                Return xml.CreateNavigator.SelectSingleNode(path).Value
            Catch exNull As NullReferenceException
                Return String.Empty
            End Try

        End Function

        ''' <summary>
        ''' Deserializes from XML file.
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="fileName">Name of the file.</param>
        ''' <returns></returns>
        Public Function DeserializeFromXmlFile(Of T)(ByVal fileName As String) As T

            Return Deserialize(Of T)(My.Computer.FileSystem.ReadAllText(fileName))

        End Function

        ''' <summary>
        ''' Deserializes the specified XML.
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="xml">The XML.</param>
        ''' <returns>T.</returns>
        ''' <exception cref="System.ArgumentNullException">xml</exception>
        Public Function Deserialize(Of T)(ByVal xml As String) As T
            If String.IsNullOrEmpty(xml) Then
                Throw New ArgumentNullException(NameOf(xml))
            End If

            Using reader = New StringReader(xml)
                Dim xmlReader = New XmlTextReader(reader)
                Return DirectCast(New XmlSerializer(GetType(T)).Deserialize(xmlReader), T)
            End Using
        End Function

        ''' <summary>
        ''' Serializes obj to XML file.
        ''' </summary>
        ''' <param name="obj">The obj.</param>
        ''' <param name="fileName">Name of the file.</param>
        Public Sub SerializeToXmlFile(ByVal obj As Object, ByVal fileName As String)

            My.Computer.FileSystem.WriteAllText(fileName, Serialize(obj), False)

        End Sub

        ''' <summary>
        ''' Serializes the specified obj to xml.
        ''' </summary>
        ''' <param name="obj">The obj.</param>
        ''' <returns></returns>
        Public Function Serialize(ByVal obj As Object) As String
            If obj Is Nothing Then
                Throw New ArgumentNullException(NameOf(obj))
            End If

            Using writer = New StringWriter()
                Using xmlWriter = New XmlTextWriter(writer)
                    Dim serilizer As New XmlSerializer(obj.[GetType]())
                    serilizer.Serialize(xmlWriter, obj)
                    Return writer.ToString()
                End Using
            End Using

        End Function

    End Module
End Namespace

File: dotNetTips.Utility.Standard/Xml/XmlHelper.cs
// ***********************************************************************
// Assembly         : dotNetTips.Utility.Standard
// Author           : David McCarter
// Created          : 02-04-2017
//
// Last Modified By : David McCarter
// Last Modified On : 08-06-2017
// ***********************************************************************
// <copyright file="XmlHelper.cs" company="dotNetTips.com - David McCarter">
//     dotNetTips.com - David McCarter
// </copyright>
// <summary></summary>
// ***********************************************************************
using dotNetTips.Utility.Standard.OOP;
using System;
using System.IO;
using System.Xml;

namespace dotNetTips.Utility.Standard.Xml
{
    /// <summary>
    /// Class XmlHelper.
    /// </summary>
    public class XmlHelper
    {
        /// <summary>
        /// Deserializes from XML file.
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="fileName">Name of the file.</param>
        /// <returns>T.</returns>
        public T DeserializeFromXmlFile<T>(string fileName)
        {
            return Deserialize<T>(File.ReadAllText(fileName));

        }

        /// <summary>
        /// Deserializes the specified XML.
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="xml">The XML.</param>
        /// <returns>T.</returns>
        /// <exception cref="ArgumentNullException">xml</exception>
        public T Deserialize<T>(string xml)
        {
            if (string.IsNullOrEmpty(xml))
            {
                throw new ArgumentNullException(nameof(xml));
            }

            using (var reader = new StringReader(xml))
            {
                //TODO: NEED TO FIGURE THIS OUT
                //using (var xmlReader = new XmlTextReader(reader))
                //{
                //    return (T)new XmlSerializer(typeof(T)).Deserialize(xmlReader);
                //}
                return default(T);
            }
        }

        /// <summary>
        /// Serializes obj to XML file.
        /// </summary>
        /// <param name="obj">The obj.</param>
        /// <param name="fileName">Name of the file.</param>
        /// <exception cref="NotImplementedException"></exception>

        public void SerializeToXmlFile(object obj, string fileName)
        {
            //TODO: NEED TO FIGURE THIS OUT
            //  My.Computer.FileSystem.WriteAllText(fileName, Serialize(obj), false);
            throw new NotImplementedException();
        }

        /// <summary>
        /// Serializes the specified obj to xml.
        /// </summary>
        /// <param name="obj">The obj.</param>
        /// <returns>System.String.</returns>
        /// <exception cref="ArgumentNullException">obj</exception>
        public string Serialize(object obj)
        {

            Encapsulation.TryValidateParam<ArgumentNullException>(obj != null);

            if (obj == null)
            {
                throw new ArgumentNullException(nameof(obj));
            }
         
            using (var writer = new StringWriter())
            {
                //TODO: BLOG POST
                using (var xmlWriter = XmlWriter.Create(writer))
                {
                    //TODO: NEED TO FIGURE THIS OUT
                    //XmlSerializer serilizer = new XmlSerializer(obj.GetType());
                    //serilizer.Serialize(xmlWriter, obj);
                    //return writer.ToString();
                    return null;
                }
            }

        }
    }
}
